### DISCUSSION TITLE
Dear Fluid Users and Developers...

### DISCUSSION THREAD

**Discussion by MatthiasWM** [fltk/fltk#1163] [Fluid]
Dear Fluid Users and Developers,

With the release of FLTK 1.4.0, Fluid 1.4 has transitioned into maintenance mode. While no new features will be added, it will continue to receive bug fixes and updates to ensure stability.

The development of Fluid 1.5 presents an exciting opportunity to modernize and future-proof the tool. By introducing C++17 (likely) and leveraging the C++ standard library, we plan to refactor and streamline the codebase thoroughly. This effort will not only enhance Fluid itself but also reveal areas where FLTK can be improved and modernized.

After nearly 30 years, it’s time for significant updates. The goals for Fluid 1.5 include:

- Full transition from autoconf to CMake for build configuration.
- Reorganized file structure for better maintainability.
- Adoption of namespaces, new classes, no global variables.
- Full utilization of the C++ standard library to simplify and enhance functionality.
- Comprehensive code documentation for clarity and ease of contribution.
- Modular design for easy extension with new widgets and project import/export capabilities.
- Foundations for multi-language support, including Python, Rust, and potentially others.

Your feedback is crucial to this evolution. Here are a few specific questions to consider:

- Should Fluid adopt a monolithic design with a single-window interface, or retain the current multi-window approach?
- Would improved internationalization support be valuable to you?
- Are there other features or improvements you’d like to see in Fluid?

Let’s shape the future of Fluid together. Your creative ideas and input will be instrumental in guiding these changes. Looking forward to hearing from you!

**Comment by Albrecht-S** [comment]
> Should Fluid adopt a monolithic design with a single-window interface, or retain the current multi-window approach?

I'm not a regular `fluid` user but here are my thoughts:
I believe that some kind of multi-window approach is definitely useful for
1. the main widget (window) display, i.e. the one that's being designed, as it is now
2. the source display window which already has a tabbed layout
3. the rest of the current windows **might** be put into a common window

Details:
1. The designed window should be a separate window so it can be resized as needed.
2. The source window should be resizable to fit the user's needs and could be moved to a separate screen by the user.
3. The 3rd window, let's call it tool window or toolbox, could hold the widget bin at the top (can be switched on/off), the widget tree at the left side (always on), a "properties" area at the right side (as needed), etc.. Is there more than that?

However, that's only an idea of a non-regular fluid user, "real" fluid users might have better ideas.

Technical note: on the Wayland platform window placement is a problem (you can't save and restore window positions), hence keeping more tools in a common window could be helpful.

**Reply by Albrecht-S** [reply]
PS: the design of fluid should not assume a large screen or multiple screens so usage of notebooks with smaller displays can be fully supported.

**Reply by MatthiasWM** [reply]
Wayland is the main reason to go monolithic TBH. We may get a nice docking tool window out of this for FLTK. Also, by making the project window part of a scrollable work surface, you can design windows larger than the current screen, which is not possible ATM.

**Comment by MoAlyousef** [comment]
I’m excited to see the upcoming improvements in fluid.
Regarding the single window vs multiwindow, apparently most designer tools come as single window these days. There was an article also noting how wayland for example is not really amenable to multiwindow apps.

The only other gui designer I had used was Qt designer via Qt creator, which comes as a single window, and I found the experience nice.

Another suggestion is using a serialization format like json or xml (via a third party library), this would make translating design files easier for other libraries and languages.

Best,
Mohammed

Sent from my iPhone

On 10 Dec 2024, at 15:59, Albrecht Schlosser ***@***.***> wrote:

﻿

Should Fluid adopt a monolithic design with a single-window interface, or retain the current multi-window approach?

I'm not a regular fluid user but here are my thoughts:
I believe that some kind of multi-window approach is definitely useful for

  1.  the main widget (window) display, i.e. the one that's being designed, as it is now
  2.  the source display window which already has a tabbed layout
  3.  the rest of the current windows might be put into a common window

Details:

  1.  The designed window should be a separate window so it can be resized as needed.
  2.  The source window should be resizable to fit the user's needs and could be on a separate screen.
  3.  The 3rd window, let's call it tool window or toolbox, could hold the widget bin at the top (can be switched on/off), the widget tree at the left side (always on), a "properties" area at the right side (as needed), etc.. Is there more than that?

However, that's only an idea of a non-regular fluid user, "real" fluid users might have better ideas.

Technical note: on the Wayland platform window placement is a problem (you can't save and restore window positions), hence keeping more tools in a common window could be helpful.

**Reply by MatthiasWM** [reply]
Wayland is what finally tipped my opinion. Fluid is unusable under Wayland unless you leave it running during the entire development phase and never reboot ;-) . I tried Qt Creator. Xcode also uses a single window for UI design.

I have been looking into XML and JSON, and I think that one of those is the way to go. The 1.4 format is mostly back compatible to 1.3, and at a first glance, this seems great, but it also generates subtle incompatibilities. Going XML would allow us to give better error messages and use many external tools. YAML is also an option. Any preference?

**Reply by MatthiasWM** [reply]
ChatGPT recommends:

For Fluid, JSON might be the best choice if:
- You value lightweight, widely adopted formats.
- You expect most project files to be programmatically generated and edited.
- You want robust library support for parsing and serialization in C++ and other languages.

I should ask ChatGPT to implement that for me (and it would probably do it, too).

**Reply by Albrecht-S** [reply]
> ChatGPT recommends

Curious: what was the question?

**Reply by MatthiasWM** [reply]
> I want to fix fluid and no longer store project files in a proprietary format. Which format do you recommend? I looked at XML, YAML and JSON.

ChatGPT said:

Each format you mentioned—XML, YAML, and JSON—has pros and cons, so the choice depends on your priorities. Here's a breakdown:

## XML
### Pros:
- Highly structured and supports complex hierarchies.
- Widely used in legacy systems and has strong tool support.
- Includes built-in validation with schemas (e.g., XSD).
### Cons:
- Verbose and can be harder to read and write manually.
- Parsing tends to be slower compared to other formats.
## YAML
### Pros:
- Human-readable and concise.
- Supports comments, making it great for editable configurations.
- Handles complex structures well.
### Cons:
- Indentation sensitivity can lead to errors.
- Parsing libraries vary in quality and aren't as universally robust.

## JSON
### Pros:
- Lightweight and widely adopted.
- Easy to parse and supported by almost every modern programming language.
- Human-readable, though less so than YAML for very complex data.
### Cons:
- No support for comments, making manual edits harder.
- Can become unwieldy for deeply nested or highly structured data.

## Recommendation

For Fluid, JSON might be the best choice if:

- You value lightweight, widely adopted formats.
- You expect most project files to be programmatically generated and edited.
- You want robust library support for parsing and serialization in C++ and other languages.
If readability and manual editing by users are high priorities, YAML could be more user-friendly, especially with the possibility of adding comments.

XML is ideal if backward compatibility or integration with older systems is required, but it’s likely overkill for modern use cases unless those features are necessary.

Let me know if you want help with library recommendations for any format!

**Reply by MatthiasWM** [reply]
Note that ChatGPT remembers that I am a developer of Fluid. I don't have to explain that. It gets really creepy when you ask it to write a unit test for any random FLTK function: it will return perfectly fine, useful, and commented code for you, explains the test, and tell you how to compile and run it. For example: "Write a unit test for Fl_Group::add()." . No further context needed.

**Reply by MoAlyousef** [reply]
If you can afford to use C++20 for fluid, there's a great library called reflect-cpp which allows serialization and deserialization of structs directly to/from several formats. Other libraries require you to parse/generate each field or property.

**Reply by MatthiasWM** [reply]
Thanks for the reference to reflect-cpp. We have not discussed the C++ version for 1.5 yet. I am somewhat worried that projects that depend on FLTK may not support higher C++ versions. I was hoping for C++17, but it still needs to be decided.

**Reply by MatthiasWM** [reply]
On a side note: Fluid already links with a working XML reader which is part of the SVG reader. Not as impressive as reflect-cpp, obviously.
